src/components/timeline/TrackScreenshots.js

```diff
@@ -0,0 +1,254 @@
+    const { screenshots, thread, isMakingPreviewSelection, width } = this.props;
+    if (isMakingPreviewSelection || offsetX === null || pageX === null) {
+      return null;
+    }
+    const screenshotIndex = this.findScreenshotAtMouse(offsetX);
+    if (screenshotIndex === null) {
+      return null;
+    }
+    const { url, windowWidth, windowHeight } = screenshots.data[
+      screenshotIndex
+    ];
+
+    // Compute the hover image's thumbnail size.
+    let hoverHeight = HOVER_HEIGHT;
+    let hoverWidth = HOVER_HEIGHT / windowHeight * windowWidth;
+
+    if (hoverWidth > HOVER_HEIGHT * HOVER_MAX_WIDTH_RATIO) {
+      // This is a really wide image, limit the height so it lays out reasonably.
+      hoverWidth = HOVER_HEIGHT * HOVER_MAX_WIDTH_RATIO;
+      hoverHeight = hoverWidth / windowWidth * windowHeight;
+    }
+
+    // Set the top so it centers around the track.
+    const top = containerTop + (TRACK_HEIGHT - hoverHeight) * 0.5;
+    const left =
+      offsetX + hoverWidth * 0.5 > width
+        ? // Stick the hover image on to the right side of the container.
+          pageX - offsetX + width - hoverWidth * 0.5
+        : // Center the hover image around the mouse.
+          pageX;
+
+    return createPortal(
+      <div className="timelineTrackScreenshotHover" style={{ left, top }}>
+        <img
+          className="timelineTrackScreenshotHoverImg"
+          src={thread.stringTable.getString(url)}
+          style={{
+            height: hoverHeight,
+            width: hoverWidth,
+          }}
+        />
+      </div>,
+      this._overlayElement
+    );
+  }
+
+  _handleMouseLeave = () => {
+    this.setState({
+      offsetX: null,
+      pageX: null,
+      containerTop: null,
+    });
+  };
+
+  _handleMouseMove = (event: SyntheticMouseEvent<HTMLDivElement>) => {
+    const { top, left } = event.currentTarget.getBoundingClientRect();
+    this.setState({
+      pageX: event.pageX,
+      offsetX: event.pageX - left,
+      containerTop: top,
+    });
+  };
+
+  render() {
+    return (
+      <div
+        className="timelineTrackScreenshot"
+        style={{ height: TRACK_HEIGHT }}
+        onMouseLeave={this._handleMouseLeave}
+        onMouseMove={this._handleMouseMove}
+      >
+        {this.renderScreenshotStrip()}
+        {this.renderHoverPreview()}
+      </div>
+    );
+  }
+}
+
+const options: ExplicitConnectOptions<OwnProps, StateProps, DispatchProps> = {
+  mapStateToProps: (state, ownProps) => {
+    const { threadIndex, screenshotId } = ownProps;
+    const selectors = selectorsForThread(threadIndex);
+    const { start, end } = getCommittedRange(state);
+    const previewSelection = getPreviewSelection(state);
+    return {
+      thread: selectors.getRangeFilteredThread(state),
+      screenshots: ensureExists(
+        selectors.getScreenshotMarkersById(state).get(screenshotId),
+        'Expected to find screenshots for the given pid'
+      ),
```

Comment:
This throws if we have no screenshot data for this id. This can happen more often than we think (eg: zooming in to a range that has no screenshot data) so I think we should handle this better.

Maybe the easiest is to return an empty array here, because the component code handles the empty array case properly:
```javascript
      screenshots: selectors.getScreenshotMarkersById(state).get(screenshotId) || EMPTY_ARRAY,
```
(of course EMPTY_ARRAY needs to be a constant for shallow equality checks).

Please also add a test for this case.

Reply:
I changed this to have a selector that always returns an empty `TracingMarker[]` when zooming in on a track.